### PR TITLE
Feature: Socks UDP Associate support

### PR DESCRIPTION
**Pull request by bearice**
as requested in  #218 , add UDP associate support to socksv5 protocol.

### CODE REVIEW

**Comment by phantomcraft** [comment]
Very good, I will test it latter.

File: src/common/socks.rs [hunk 1]
```diff
@@ -642,4 +849,36 @@ mod tests {
...
+            0, 5, //port 5
+            1, 2, 3, 4, // data
+        ]);
+        assert_eq!(encode_socks_frame(frame).unwrap(), expected)
+    }
+
+    #[test(tokio::test)]
+    async fn test_decode_frame() {
+        let input = Bytes::from_static(&[
+            5, 3, 0, 1, // ver 5 cmd 1 resv 0 type 4
```
Comment (bearice) [inline src/common/socks.rs:871]: type 4?
typo here

File: src/listeners/socks.rs [hunk 2]
```diff
@@ -119,38 +132,92 @@ impl SocksListener {
...
+                ctx.on_error(err_msg("unknown cmd")).await;
+                debug!("unknown cmd: {:?}", request.cmd);
+            }
         }
         Ok(())
     }
 }
+
 struct Callback {
+    cmd: u8,
```
Comment (bearice) [inline src/listeners/socks.rs:206]: should remove this field

**Comment by bearice** [comment]
I think it's ready to merge.
@phantomcraft can you please test this branch to see if it works with shandowsocks or any other clients/server you prefer?

**Comment by phantomcraft** [comment]
Here is the test:

**Configuration:**

```
apiVersion: v1alpha
kind: ProxyDefinition

ioParams:
  bufferSize: 65536
  useSplice: true

metrics:
  bind: "0.0.0.0:8888"

timeouts:
  idle: 10
  udp: 10

listeners:
  - name: socks
    bind: 0.0.0.0:1081
    auth:
      required: false
      users:
        - username: a
          password: a
      cmd:
        - test
        - "#USER#"
        - ==
        - "#PASS#"
      cache:
        timeout: 10

connectors:
  - name: socks
    server: 127.0.0.1:1080
    port: 1080

rules:

accessLog:
  path: access.log
  format: json

```

127.0.0.1:1080 is a Shadowsocks proxy with full UDP connectivity.

When using a secondary program to redirect the listener 0.0.0.0:1081 to a UDP TPROXY:

```
user@transns-fortress_ID=[j3SL19WL0Q]:~$ dig @1.1.1.1 t.co
;; communications error to 1.1.1.1#53: timed out
;; communications error to 1.1.1.1#53: timed out
;; communications error to 1.1.1.1#53: timed out
; <<>> DiG 9.18.7-1-Debian <<>> @1.1.1.1 t.co
; (1 server found)
;; global options: +cmd
;; no servers could be reached
```

TCP also doesn't work when redirecting 0.0.0.0:1081 to a TCP Redir port:

```
user@transns-fortress_ID=[j3SL19WL0Q]:~$ curl 1.1.1.1:80
curl: (56) Recv failure: Connection reset by peer
```

redproxy-rs output log:

```
user@localhost:/dev/shm/redproxy-rs-feat-socks-udp-assoc$ target/release/redproxy-rs --config /dev/shm/config.yaml 
[2022-10-24T18:51:55Z INFO  redproxy_rs::listeners::socks] socks listening on 0.0.0.0:1081
[2022-10-24T18:51:55Z INFO  redproxy_rs::metrics] metrics server listening on 0.0.0.0:8888
[2022-10-24T18:52:00Z INFO  redproxy_rs] implicitly denied: id=0 l=socks c=<null> s=127.0.0.1:51469 t=1.1.1.1:80 f=TcpForward
[2022-10-24T18:52:04Z INFO  redproxy_rs] implicitly denied: id=1 l=socks c=<null> s=127.0.0.1:54507 t=0.0.0.0:0 f=UdpForward
[2022-10-24T18:52:09Z INFO  redproxy_rs] implicitly denied: id=2 l=socks c=<null> s=127.0.0.1:48075 t=0.0.0.0:0 f=UdpForward
[2022-10-24T18:52:14Z INFO  redproxy_rs] implicitly denied: id=3 l=socks c=<null> s=127.0.0.1:39297 t=0.0.0.0:0 f=UdpForward
```

I sent to your e-mail bearice@... the script that uses the secondary program that I use to connect to Shadowsocks.

**Comment by bearice** [comment]
you should have at least one rule entry point to the default connector, or all connections will be denied by default;
refs: https://github.com/MengJiangProject/redproxy-rs/blob/master/config.yaml#L146
example:
```yaml
...
rules:
  - target: direct
 ...
```

**Comment by bearice** [comment]
I tested it locallly with a v2ray/vmess setup and it works, but not tested with other softwares yet.

**Comment by phantomcraft** [comment]
Config:

```
apiVersion: v1alpha
kind: ProxyDefinition

ioParams:
  bufferSize: 65536
  useSplice: true

metrics:
  bind: "0.0.0.0:8888"

timeouts:
  idle: 10
  udp: 10

listeners:
  - name: socks
    bind: 0.0.0.0:1082
    auth:
      required: false
      users:
        - username: a
          password: a
      cmd:
        - test
        - "#USER#"
        - ==
        - "#PASS#"
      cache:
        timeout: 10

connectors:
  - name: socks
    server: 127.0.0.1:1080
    port: 1080
  - name: direct
rules:
  - target: direct

accessLog:
  path: access.log
  format: json
```

Results:

```
user@transns-fortress_ID=[QJHNZEdZuB]:~$ time dig @114.114.114.114 g.co

; <<>> DiG 9.18.7-1-Debian <<>> @114.114.114.114 g.co
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 22042
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;g.co.				IN	A

;; ANSWER SECTION:
g.co.			178	IN	A	142.251.209.14

;; Query time: 307 msec
;; SERVER: 114.114.114.114#53(114.114.114.114) (UDP)
;; WHEN: Tue Oct 25 16:14:45 EDT 2022
;; MSG SIZE  rcvd: 49


real	0m0.330s
user	0m0.008s
sys	0m0.004s
user@transns-fortress_ID=[QJHNZEdZuB]:~$ time dig @1.1.1.1 g.co

; <<>> DiG 9.18.7-1-Debian <<>> @1.1.1.1 g.co
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 46063
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;g.co.				IN	A

;; ANSWER SECTION:
g.co.			11	IN	A	142.250.179.206

;; Query time: 229 msec
;; SERVER: 1.1.1.1#53(1.1.1.1) (UDP)
;; WHEN: Tue Oct 25 16:14:50 EDT 2022
;; MSG SIZE  rcvd: 49


real	0m0.250s
user	0m0.008s
sys	0m0.002s
user@transns-fortress_ID=[QJHNZEdZuB]:~$ curl 1.1.1.1:80
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
```

But the connections are not going through socks5@127.0.0.1:1080, but by direct.

**Comment by bearice** [comment]
it goes to 'direct' because you told it to do so:
```yaml
rules:
  - target: direct
```
this means direct all traffic into 'direct' connector.
change to 'socks' and try.

**Comment by phantomcraft** [comment]
```
user@transns-fortress_ID=[jjoqrFfjlt]:~$ time dig @9.9.9.9 g.co
;; communications error to 9.9.9.9#53: timed out
;; communications error to 9.9.9.9#53: timed out
;; communications error to 9.9.9.9#53: timed out

; <<>> DiG 9.18.7-1-Debian <<>> @9.9.9.9 g.co
; (1 server found)
;; global options: +cmd
;; no servers could be reached


real	0m15.033s
user	0m0.008s
sys	0m0.004s
user@transns-fortress_ID=[jjoqrFfjlt]:~$ time dig @2606:4700:4700::1111 g.co
;; communications error to 2606:4700:4700::1111#53: timed out
;; communications error to 2606:4700:4700::1111#53: timed out
;; communications error to 2606:4700:4700::1111#53: timed out

; <<>> DiG 9.18.7-1-Debian <<>> @2606:4700:4700::1111 g.co
; (1 server found)
;; global options: +cmd
;; no servers could be reached


real	0m15.034s
user	0m0.006s
sys	0m0.006s
```

```
user@localhost:~$ /dev/shm/redproxy-rs-feat-socks-udp-assoc/target/release/redproxy-rs --log debug --config /dev/shm/config.yaml 
[2022-10-26T02:03:18Z INFO  redproxy_rs::listeners::socks] socks listening on 0.0.0.0:1082
[2022-10-26T02:03:18Z INFO  redproxy_rs::metrics] metrics server listening on 0.0.0.0:8888
[2022-10-26T02:03:24Z DEBUG redproxy_rs::listeners::socks] socks: connected from 127.0.0.1:58025
[2022-10-26T02:03:24Z DEBUG redproxy_rs::listeners::socks] request SocksRequest { version: 5, cmd: 3, target: SocketAddr(9.9.9.9:53), auth: None }
[2022-10-26T02:03:24Z DEBUG redproxy_rs::context] set_state: ctx=id=0 l=socks c=<null> s=127.0.0.1:58025 t=0.0.0.0:0 f=UdpForward state=ClientRequested
[2022-10-26T02:03:24Z WARN  redproxy_rs] failed to connect to upstream: unsupported connector feature: UdpForward (src/main.rs:201:17) 
    cause: None 
    ctx: id=0 l=socks c=<null> s=127.0.0.1:58025 t=0.0.0.0:0 f=UdpForward
[2022-10-26T02:03:24Z DEBUG redproxy_rs::context] set_state: ctx=id=0 l=socks c=<null> s=127.0.0.1:58025 t=0.0.0.0:0 f=UdpForward state=ErrorOccured
[2022-10-26T02:03:29Z DEBUG redproxy_rs::listeners::socks] socks: connected from 127.0.0.1:43063
[2022-10-26T02:03:29Z DEBUG redproxy_rs::listeners::socks] request SocksRequest { version: 5, cmd: 3, target: SocketAddr(9.9.9.9:53), auth: None }
[2022-10-26T02:03:29Z DEBUG redproxy_rs::context] set_state: ctx=id=1 l=socks c=<null> s=127.0.0.1:43063 t=0.0.0.0:0 f=UdpForward state=ClientRequested
[2022-10-26T02:03:29Z WARN  redproxy_rs] failed to connect to upstream: unsupported connector feature: UdpForward (src/main.rs:201:17) 
    cause: None 
    ctx: id=1 l=socks c=<null> s=127.0.0.1:43063 t=0.0.0.0:0 f=UdpForward
[2022-10-26T02:03:29Z DEBUG redproxy_rs::context] set_state: ctx=id=1 l=socks c=<null> s=127.0.0.1:43063 t=0.0.0.0:0 f=UdpForward state=ErrorOccured
[2022-10-26T02:03:34Z DEBUG redproxy_rs::listeners::socks] socks: connected from 127.0.0.1:60179
[2022-10-26T02:03:34Z DEBUG redproxy_rs::listeners::socks] request SocksRequest { version: 5, cmd: 3, target: SocketAddr(9.9.9.9:53), auth: None }
[2022-10-26T02:03:34Z DEBUG redproxy_rs::context] set_state: ctx=id=2 l=socks c=<null> s=127.0.0.1:60179 t=0.0.0.0:0 f=UdpForward state=ClientRequested
[2022-10-26T02:03:34Z WARN  redproxy_rs] failed to connect to upstream: unsupported connector feature: UdpForward (src/main.rs:201:17) 
    cause: None 
    ctx: id=2 l=socks c=<null> s=127.0.0.1:60179 t=0.0.0.0:0 f=UdpForward
[2022-10-26T02:03:34Z DEBUG redproxy_rs::context] set_state: ctx=id=2 l=socks c=<null> s=127.0.0.1:60179 t=0.0.0.0:0 f=UdpForward state=ErrorOccured
[2022-10-26T02:03:42Z DEBUG redproxy_rs::listeners::socks] socks: connected from 127.0.0.1:41677
[2022-10-26T02:03:42Z DEBUG redproxy_rs::listeners::socks] request SocksRequest { version: 5, cmd: 3, target: SocketAddr([2606:4700:4700::1111]:53), auth: None }
[2022-10-26T02:03:42Z DEBUG redproxy_rs::context] set_state: ctx=id=3 l=socks c=<null> s=127.0.0.1:41677 t=0.0.0.0:0 f=UdpForward state=ClientRequested
[2022-10-26T02:03:42Z WARN  redproxy_rs] failed to connect to upstream: unsupported connector feature: UdpForward (src/main.rs:201:17) 
    cause: None 
    ctx: id=3 l=socks c=<null> s=127.0.0.1:41677 t=0.0.0.0:0 f=UdpForward
[2022-10-26T02:03:42Z DEBUG redproxy_rs::context] set_state: ctx=id=3 l=socks c=<null> s=127.0.0.1:41677 t=0.0.0.0:0 f=UdpForward state=ErrorOccured
```

Config used:

```
apiVersion: v1alpha
kind: ProxyDefinition

ioParams:
  bufferSize: 65536
  useSplice: true

metrics:
  bind: "0.0.0.0:8888"

timeouts:
  idle: 10
  udp: 10

listeners:
  - name: socks
    bind: 0.0.0.0:1082
    auth:
      required: false
      users:
        - username: a
          password: a
      cmd:
        - test
        - "#USER#"
        - ==
        - "#PASS#"
      cache:
        timeout: 10

connectors:
  - name: socks
    server: 127.0.0.1
    port: 1080

rules:
  - target: socks

accessLog:
  path: access.log
  format: json
```

I used the script I sent to your e-mail. It work with any other proxy forwarder.

**Comment by bearice** [comment]
All right, thanks for your feedback, let me do some more research.

**Comment by bearice** [comment]
@phantomcraft I think I have fixed this problem and tested it locally, can you run your tests again to confirm this? Thanks.

**Comment by phantomcraft** [comment]
```
user@transns-fortress_ID=[4IIGiahj1Y]:~$ dig @1.1.1.1 g.co

; <<>> DiG 9.18.7-1-Debian <<>> @1.1.1.1 g.co
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 63533
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;g.co.				IN	A

;; ANSWER SECTION:
g.co.			75	IN	A	216.58.212.174

;; Query time: 669 msec
;; SERVER: 1.1.1.1#53(1.1.1.1) (UDP)
;; WHEN: Thu Oct 27 16:01:00 EDT 2022
;; MSG SIZE  rcvd: 49

user@transns-fortress_ID=[4IIGiahj1Y]:~$ dig @2606:4700:4700::1001 g.co

; <<>> DiG 9.18.7-1-Debian <<>> @2606:4700:4700::1001 g.co
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 43422
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;g.co.				IN	A

;; ANSWER SECTION:
g.co.			234	IN	A	142.250.186.110

;; Query time: 719 msec
;; SERVER: 2606:4700:4700::1001#53(2606:4700:4700::1001) (UDP)
;; WHEN: Thu Oct 27 16:01:22 EDT 2022
;; MSG SIZE  rcvd: 49

user@transns-fortress_ID=[4IIGiahj1Y]:~$ curl 1.1.1.1
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>cloudflare</center>
</body>
</html>

user@transns-fortress_ID=[4kVVxkiEGT]:~$ curl http://[2a00:1450:400e:801::200e]
<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 404 (Not Found)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL <code>/</code> was not found on this server.  <ins>That’s all we know.</ins>

```

```
user@localhost:/dev/shm/redproxy-rs-feat-socks-udp-assoc$ ./target/release/redproxy-rs --log debug --config /dev/shm/config.yaml 
[2022-10-27T20:00:29Z INFO  redproxy_rs::listeners::socks] socks listening on 0.0.0.0:1082
[2022-10-27T20:00:29Z INFO  redproxy_rs::metrics] metrics server listening on 0.0.0.0:8888
[2022-10-27T20:00:59Z DEBUG redproxy_rs::listeners::socks] socks: connected from 127.0.0.1:49467
[2022-10-27T20:00:59Z DEBUG redproxy_rs::listeners::socks] request SocksRequest { version: 5, cmd: 3, target: SocketAddr(1.1.1.1:53), auth: None }
[2022-10-27T20:00:59Z DEBUG redproxy_rs::context] set_state: ctx=id=0 l=socks c=<null> s=127.0.0.1:49467 t=0.0.0.0:0 f=UdpForward state=ClientRequested
[2022-10-27T20:00:59Z DEBUG redproxy_rs::context] set_state: ctx=id=0 l=socks c=<null> s=127.0.0.1:49467 t=0.0.0.0:0 f=UdpForward state=ServerConnecting
[2022-10-27T20:00:59Z DEBUG redproxy_rs::context] set_state: ctx=id=0 l=socks c=socks s=127.0.0.1:49467 t=0.0.0.0:0 f=UdpForward state=Connected
[2022-10-27T20:01:10Z DEBUG redproxy_rs::context] set_state: ctx=id=0 l=socks c=socks s=127.0.0.1:49467 t=0.0.0.0:0 f=UdpForward state=ClientShutdown
[2022-10-27T20:01:10Z WARN  redproxy_rs] error in io thread: idle timeout (src/copy.rs:314:28) 
    cause: None 
    ctx: id=0 l=socks c=socks s=127.0.0.1:49467 t=0.0.0.0:0 f=UdpForward
[2022-10-27T20:01:10Z DEBUG redproxy_rs::context] set_state: ctx=id=0 l=socks c=socks s=127.0.0.1:49467 t=0.0.0.0:0 f=UdpForward state=ErrorOccured
[2022-10-27T20:01:21Z DEBUG redproxy_rs::listeners::socks] socks: connected from 127.0.0.1:38751
[2022-10-27T20:01:21Z DEBUG redproxy_rs::listeners::socks] request SocksRequest { version: 5, cmd: 3, target: SocketAddr([2606:4700:4700::1001]:53), auth: None }
[2022-10-27T20:01:21Z DEBUG redproxy_rs::context] set_state: ctx=id=1 l=socks c=<null> s=127.0.0.1:38751 t=0.0.0.0:0 f=UdpForward state=ClientRequested
[2022-10-27T20:01:21Z DEBUG redproxy_rs::context] set_state: ctx=id=1 l=socks c=<null> s=127.0.0.1:38751 t=0.0.0.0:0 f=UdpForward state=ServerConnecting
[2022-10-27T20:01:21Z DEBUG redproxy_rs::context] set_state: ctx=id=1 l=socks c=socks s=127.0.0.1:38751 t=0.0.0.0:0 f=UdpForward state=Connected
[2022-10-27T20:01:32Z DEBUG redproxy_rs::context] set_state: ctx=id=1 l=socks c=socks s=127.0.0.1:38751 t=0.0.0.0:0 f=UdpForward state=ClientShutdown
[2022-10-27T20:01:32Z WARN  redproxy_rs] error in io thread: idle timeout (src/copy.rs:314:28) 
    cause: None 
    ctx: id=1 l=socks c=socks s=127.0.0.1:38751 t=0.0.0.0:0 f=UdpForward
[2022-10-27T20:01:32Z DEBUG redproxy_rs::context] set_state: ctx=id=1 l=socks c=socks s=127.0.0.1:38751 t=0.0.0.0:0 f=UdpForward state=ErrorOccured
[2022-10-27T20:01:33Z DEBUG redproxy_rs::listeners::socks] socks: connected from 127.0.0.1:35033
[2022-10-27T20:01:33Z DEBUG redproxy_rs::listeners::socks] request SocksRequest { version: 5, cmd: 1, target: SocketAddr(1.1.1.1:80), auth: None }
[2022-10-27T20:01:33Z DEBUG redproxy_rs::context] set_state: ctx=id=2 l=socks c=<null> s=127.0.0.1:35033 t=1.1.1.1:80 f=TcpForward state=ClientRequested
[2022-10-27T20:01:33Z DEBUG redproxy_rs::context] set_state: ctx=id=2 l=socks c=<null> s=127.0.0.1:35033 t=1.1.1.1:80 f=TcpForward state=ServerConnecting
[2022-10-27T20:01:33Z DEBUG redproxy_rs::context] set_state: ctx=id=2 l=socks c=socks s=127.0.0.1:35033 t=1.1.1.1:80 f=TcpForward state=Connected
[2022-10-27T20:01:39Z DEBUG redproxy_rs::context] set_state: ctx=id=2 l=socks c=socks s=127.0.0.1:35033 t=1.1.1.1:80 f=TcpForward state=ClientShutdown
[2022-10-27T20:05:10Z DEBUG redproxy_rs::listeners::socks] socks: connected from 127.0.0.1:49991
[2022-10-27T20:05:10Z DEBUG redproxy_rs::listeners::socks] request SocksRequest { version: 5, cmd: 1, target: SocketAddr([2a00:1450:400e:801::200e]:80), auth: None }
[2022-10-27T20:05:10Z DEBUG redproxy_rs::context] set_state: ctx=id=3 l=socks c=<null> s=127.0.0.1:49991 t=[2a00:1450:400e:801::200e]:80 f=TcpForward state=ClientRequested
[2022-10-27T20:05:10Z DEBUG redproxy_rs::context] set_state: ctx=id=3 l=socks c=<null> s=127.0.0.1:49991 t=[2a00:1450:400e:801::200e]:80 f=TcpForward state=ServerConnecting
[2022-10-27T20:05:10Z DEBUG redproxy_rs::context] set_state: ctx=id=3 l=socks c=socks s=127.0.0.1:49991 t=[2a00:1450:400e:801::200e]:80 f=TcpForward state=Connected
[2022-10-27T20:05:16Z DEBUG redproxy_rs::context] set_state: ctx=id=3 l=socks c=socks s=127.0.0.1:49991 t=[2a00:1450:400e:801::200e]:80 f=TcpForward state=ClientShutdown
```

Now it's fully working, I think it's ready to merge.

**Comment by phantomcraft** [comment]
For making SOCKS5 protocol fully functional, the next step would be adding BIND support, it's just a suggestion.

**Pull request merged**